### ISSUE TITLE
[Questions] DALI reader execution, timing the read calls & S3 integration

### ISSUE THREAD

**Issue by sohomb91** [NVIDIA/DALI#4269]
#### Sample Code Reference:
```
@pipeline_def
def preImg_pipe(self):
        start = time.monotonic()
        imgs, labels = fn.readers.file(file_root=dirs[0],
                                       random_shuffle=True,
                                       initial_fill=3670,
                                       pad_last_batch=True,
                                       name='Reader')
        read_time = time.monotonic() - start

        imgs = fn.decoders.image(imgs, device='cpu')
        imgs = fn.color_space_conversion(imgs, image_type=types.RGB, output_type=types.GRAY)
        angle = fn.random.uniform(range=(-10.0, 10.0))
        imgs = fn.rotate(imgs, angle=angle, fill_value=0)
        imgs = fn.resize(imgs,
                         device='cpu',
                         resize_x=32,
                         resize_y=32,
                         mode='stretch')

        return imgs, labels, read_time
		
data_pipe = preImg_pipe(batch_size = 100, num_threads=10, device_id=None)
data_pipe.build()

train_dataloader = DALIGenericIterator([data_pipe],
                                           output_map=['images', 'labels', 'read_times'],
                                           reader_name="Reader",
                                           last_batch_policy=LastBatchPolicy.PARTIAL)
								   							   
for i, data in enumerate(train_loader):
	# Extract images, labels to feed to model & read_times to update the metrics report
```
Based on the above, I have a few queries:

#### 1. What is the order of execution of "File read"?
As evident, when calling `enumerate(train_loader)`, DALI execution goes back to the built pipeline graph and executes the list of operations defined within the graph. 
So when it executes `fn.readers.file`, I want to understand the internal order of the file reads: 
Does it:
1. List all the files at once and then reads all of them? Kind of like ***List-all --> Read-all*** format? or
2. List one then read one, then move forward? Like ***List-1st --> Read-1st --> List 2nd --> Read 2nd --> ...*** format?

#### 2. Best way to extract the time taken during file reads?
This is somewhat attached to the previous query...
I am doing some performance testing on some of available storage devices with AI workload, so I need to measure the time taken for just the image reads. As seen in above code I am measuring the time as follows:
```
start = time.monotonic()
imgs, labels = fn.readers.file(file_root=dirs[0],
                                       random_shuffle=True,
                                       initial_fill=3670,
                                       pad_last_batch=True,
                                       name='Reader')
read_time = time.monotonic() - start
```
   But I guess the above is a pretty crude way of doing it -- as it would also encompass the "listing time" as well -- which I want to 
   avoid and granularize it to just have the "read time"! Another problem with this approach, might be related with parallelism 
   `num_threads=10` -- with 10 threads running in parallel, all 10 running a file read each at the same time, I am not sure if I am having 
   the exact timings with this implementation.
   Is there a way I can extract the time taken to read an entire batch (encompassing the 10 thread parallelism)? 

#### 3. Process of reading the dataset from AWS S3?
As said, I am doing some performance testing on some of the storage devices -- This brings me to the current query of how we could integrate calls to AWS S3, to read the image files from a specific bucket? -- Is there any integration of DALI with the `boto/boto3` packages to enable calls to AWS or similar SDK's for Azure?

**Comment by JanuszL** [comment]
Hi @sohomb91,

Thank you for reaching out. Let me answer all your questions one by one:
1) >As evident, when calling enumerate(train_loader), DALI execution goes back to the built pipeline graph and executes the list of operations defined within the graph.

It happens only when the pipeline is built. Then the operators are instantiated on the native DALI side and no more Python code related to the graph definition is executed. However, there is [a debug mode](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/debug_mode.html) that eagerly executes the DALI pipeline. It is meant for debugging, not for performance runs.

> So when it executes fn.readers.file, I want to understand the internal order of the file reads:

DALI discovers all files and assigns labels to them. If the shuffling is enabled the shuffling buffer size number of files is read one by one, and then each iteration batch size number of files is read in addition.

2) >Best way to extract the time taken during file reads?

I recommend capturing [a Nsigh System](https://developer.nvidia.com/nsight-systems) profile, like this one:
![image](https://user-images.githubusercontent.com/39967756/191464188-7490b971-4786-4032-8da5-a35b9db1d9dd.png)

So reading is done by a separate prefetching thread that either reads files directly or mmap opened file, and the actual page fault and access to the data happens during operator execution (which is also in many cases asynchronous to the main python thread).
The number of DALI threads doesn't impact the reading speed.

3) > Process of reading the dataset from AWS S3?

I recommend using Python defined reading pattern vis [the `external_source` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html) (check `parallel` and `prefetch_queue_depth` parameters for the optimal performance).

**Comment by sohomb91** [comment]
Hi @JanuszL ,
Many thanks for your prompt response on this... :)

> I recommend using Python defined reading pattern vis [the external_source operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html) (check parallel and prefetch_queue_depth parameters for the optimal performance).

Got it... might make sense...
I will give this a read and try to implement a standalone code to work this out. If I do face any issues, I will reach out! 


> I recommend capturing [a Nsigh System](https://developer.nvidia.com/nsight-systems) profile, like this one:

This is quite a good approach... I was looking exactly for something like this that could have encompassed all the performance metrics for my application's algo. I guess, I will have to utilize the CLI tool to get the metrics as mostly I will have scripts to fetch and analyze them. 
Many thanks for suggesting this.

**Comment by sohomb91** [comment]
> It happens only when the pipeline is built. Then the operators are instantiated on the native DALI side and no more Python code related to the graph definition is executed. However, there is [a debug mode](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/debug_mode.html) that eagerly executes the DALI pipeline. It is meant for debugging, not for performance runs.

Ohh, in that case, I think my understanding might have been off....
So you mean when we build the pipeline, the individual operators already work on the data (read, resize, augment, etc), and store them to memory -- which is then just fetched when we call the pipeline.run() or wrap it in `DALIGenericIterator`?

**Comment by JanuszL** [comment]
> Ohh, in that case, I think my understanding might have been off.... So you mean when we build the pipeline, the individual operators already work on the data (read, resize, augment, etc), and store them to memory -- which is then just fetched when we call the pipeline.run() or wrap it in `DALIGenericIterator`?

I would put it in a different way. DALI is a native library with a Python interface. It works asynchronously with the main Python thread. The Python API serves to define the processing flow (instantiate operators and connect them), schedule pipeline execution, and obtain the results. The part you timed is just an instantiation of the operator. The actual execution happens asynchronously at the native level and can't be easily measured from the python side.

**Comment by sohomb91** [comment]
Ahh ok... I got that now.
So, in any case, I guess the `Nsight` could be the best option, as you suggested!

Thanks for all the help, I'll explore that further. :)
